File: cowork-roadmap/src/main/java/com/cowork/roadmap/global/team/ProjectionCatchUpPolicy.java
package com.cowork.roadmap.global.team;

import java.util.Map;
import java.util.Set;

record ProjectionCheckpoint(long nextOffset, String topicId, Long invalidCheckpointOffset, Long snapshotCompletedOffset,
        Long invalidRecordOffset) {

    ProjectionCheckpoint(long nextOffset, String topicId, Long invalidCheckpointOffset, Long snapshotCompletedOffset) {
        this(nextOffset, topicId, invalidCheckpointOffset, snapshotCompletedOffset, null);
    }
}

record ProjectionBarrier(String topicId, long targetOffset) {
}

record ProjectionPartitionRange(int partition, long beginningOffset, long endOffset) {
}

record ProjectionBrokerRange(long beginningOffset, long endOffset) {
}

record ProjectionTopicState(String topicId, Map<Integer, ProjectionBrokerRange> ranges) {
}

record ProjectionSeekDecision(long seekOffset, String topicId, Long invalidCheckpointOffset) {
}

final class ProjectionCatchUpPolicy {

    private ProjectionCatchUpPolicy() {
    }

    static ProjectionSeekDecision seekDecision(ProjectionCheckpoint checkpoint,
            long beginningOffset,
            long endOffset,
            String currentTopicId) {
        if (checkpoint == null) {
            return new ProjectionSeekDecision(beginningOffset, currentTopicId, null);
        }
        if (!checkpoint.topicId().equals(currentTopicId)) {
            return new ProjectionSeekDecision(beginningOffset, currentTopicId, checkpoint.nextOffset());
        }
        if (checkpoint.invalidCheckpointOffset() != null) {
            long safeOffset = checkpoint.nextOffset() >= beginningOffset && checkpoint.nextOffset() <= endOffset
                    ? checkpoint.nextOffset()
                    : beginningOffset;
            return new ProjectionSeekDecision(safeOffset, currentTopicId, checkpoint.invalidCheckpointOffset());
        }
        if (checkpoint.nextOffset() < beginningOffset || checkpoint.nextOffset() > endOffset) {
            return new ProjectionSeekDecision(beginningOffset, currentTopicId, checkpoint.nextOffset());
        }
        return new ProjectionSeekDecision(checkpoint.nextOffset(), currentTopicId, null);
    }

    static boolean isReady(Set<Integer> brokerPartitions,
            String currentTopicId,
            Map<Integer, ProjectionBarrier> barriers,
            Map<Integer, ProjectionCheckpoint> checkpoints,
            Map<Integer, ProjectionBrokerRange> brokerRanges) {
        if (brokerPartitions.isEmpty() || !barriers.keySet().equals(brokerPartitions)
                || !checkpoints.keySet().equals(brokerPartitions) || !brokerRanges.keySet().equals(brokerPartitions)) {
            return false;
        }
        for (Integer partition : brokerPartitions) {
            ProjectionBarrier barrier = barriers.get(partition);
            ProjectionCheckpoint checkpoint = checkpoints.get(partition);
            ProjectionBrokerRange range = brokerRanges.get(partition);
            if (!barrier.topicId().equals(currentTopicId) || !checkpoint.topicId().equals(currentTopicId)
                    || checkpoint.invalidCheckpointOffset() != null || checkpoint.snapshotCompletedOffset() == null
                    || checkpoint.invalidRecordOffset() != null
                    || checkpoint.nextOffset() <= checkpoint.snapshotCompletedOffset()
                    || checkpoint.nextOffset() < barrier.targetOffset() || barrier.targetOffset() > range.endOffset()
                    || checkpoint.snapshotCompletedOffset() < range.beginningOffset()
                    || checkpoint.nextOffset() < range.beginningOffset()
                    || checkpoint.nextOffset() != range.endOffset()) {
                return false;
            }
        }
        return true;
    }
}

File: cowork-channel/src/main/kotlin/com/cowork/channel/global/projection/ProjectionCatchUpPolicy.kt
package com.cowork.channel.global.projection

data class ProjectionStream(val consumerGroup: String, val topic: String, val expectedSource: String?)

data class ProjectionCheckpoint(
    val nextOffset: Long,
    val topicId: String,
    val invalidCheckpointOffset: Long? = null,
    val snapshotCompletedOffset: Long? = null,
    val invalidRecordOffset: Long? = null,
)

data class ProjectionBarrier(val topicId: String, val targetOffset: Long)

data class ProjectionPartitionRange(val partition: Int, val beginningOffset: Long, val endOffset: Long)

data class ProjectionBrokerRange(val beginningOffset: Long, val endOffset: Long)

data class ProjectionTopicState(val topicId: String, val ranges: Map<Int, ProjectionBrokerRange>)

data class ProjectionSeekDecision(val seekOffset: Long, val topicId: String, val invalidCheckpointOffset: Long?)

object ProjectionCatchUpPolicy {
    fun seekDecision(
        checkpoint: ProjectionCheckpoint?,
        beginningOffset: Long,
        endOffset: Long,
        currentTopicId: String,
    ): ProjectionSeekDecision {
        if (checkpoint == null) return ProjectionSeekDecision(beginningOffset, currentTopicId, null)
        if (checkpoint.topicId != currentTopicId) {
            return ProjectionSeekDecision(beginningOffset, currentTopicId, checkpoint.nextOffset)
        }
        if (checkpoint.invalidCheckpointOffset != null) {
            val safeOffset = checkpoint.nextOffset.takeIf { it in beginningOffset..endOffset } ?: beginningOffset
            return ProjectionSeekDecision(safeOffset, currentTopicId, checkpoint.invalidCheckpointOffset)
        }
        if (checkpoint.nextOffset < beginningOffset || checkpoint.nextOffset > endOffset) {
            return ProjectionSeekDecision(beginningOffset, currentTopicId, checkpoint.nextOffset)
        }
        return ProjectionSeekDecision(checkpoint.nextOffset, currentTopicId, null)
    }

    fun isReady(
        requiredStreams: Set<ProjectionStream>,
        barriers: Map<ProjectionStream, Map<Int, ProjectionBarrier>>,
        checkpoints: Map<ProjectionStream, Map<Int, ProjectionCheckpoint>>,
    ): Boolean = requiredStreams.all { stream ->
        val streamBarriers = barriers[stream].orEmpty()
        val streamCheckpoints = checkpoints[stream].orEmpty()
        streamBarriers.isNotEmpty() &&
            streamBarriers.all { (partition, barrier) ->
                val checkpoint = streamCheckpoints[partition]
                checkpoint != null &&
                    checkpoint.topicId == barrier.topicId &&
                    checkpoint.invalidCheckpointOffset == null &&
                    checkpoint.invalidRecordOffset == null &&
                    checkpoint.snapshotCompletedOffset != null &&
                    checkpoint.nextOffset > checkpoint.snapshotCompletedOffset &&
                    checkpoint.nextOffset >= barrier.targetOffset
            }
    }

    fun matchesBrokerState(
        requiredStreams: Set<ProjectionStream>,
        barriers: Map<ProjectionStream, Map<Int, ProjectionBarrier>>,
        checkpoints: Map<ProjectionStream, Map<Int, ProjectionCheckpoint>>,
        topicStates: Map<String, ProjectionTopicState>,
    ): Boolean = requiredStreams.all { stream ->
        val topicState = topicStates[stream.topic] ?: return@all false
        val streamBarriers = barriers[stream].orEmpty()
        val streamCheckpoints = checkpoints[stream].orEmpty()
        val brokerPartitions = topicState.ranges.keys
        if (streamBarriers.keys != brokerPartitions || streamCheckpoints.keys != brokerPartitions) return@all false

        topicState.ranges.all { (partition, range) ->
            val barrier = streamBarriers.getValue(partition)
            val checkpoint = streamCheckpoints.getValue(partition)
            barrier.topicId == topicState.topicId &&
                checkpoint.topicId == topicState.topicId &&
                checkpoint.snapshotCompletedOffset?.let { it >= range.beginningOffset } == true &&
                barrier.targetOffset <= range.endOffset &&
                checkpoint.nextOffset == range.endOffset
        }
    }
}
